File: 0001-two-sum/solution.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	for n := 0; n < len(nums)-1; n++ {
		for m := 1; m < len(nums); m++ {
			if nums[n]+nums[m] == target {
				fmt.Println(nums[n], nums[m])
				return []int{n, m}
			}
		}
	}
	return []int{}
}

func twoSum2(nums []int, target int) []int {
	hashMap := make(map[int]int)

	for i, num := range nums {
		complement := target - num

		if index, found := hashMap[complement]; found {
			return []int{index, i}
		}
		hashMap[num] = i
	}

	return []int{}
}

func main() {
	num := []int{2, 7, 11, 15}
	target := 9

	result := twoSum(num, target)
	result2 := twoSum2(num, target)
	fmt.Println(result)
	fmt.Println(result2)
}

File: 0001-two-sum/solution.swift
func twoSum_1(_ nums: [Int], _ target: Int) -> [Int] {
    for n in 0..<nums.count-1 {
        for m in 1..<nums.count {
            if nums[n] + nums[m] == target {
                print(nums[n], nums[m]) return [n, m]
            }
        }
    }
    return []
}
//Time Complexity O(n^2) - The algorithm uses two nested loops to iterate through the array, resulting in a quadratic time complexity.
//Space Complexity O(1) - The algorithm uses a constant amount of extra space, regardless of the input size.


func twoSum_best(_ nums: [Int], _ target: Int) -> [Int] {
    var hashMap = [Int: Int]()
    
    for (index, num) in nums.enumerated() {
        let complement = target - num
        if let complementIndex = hashMap[complement] {
            print(nums[complementIndex], nums[index])
            return [complementIndex, index]
        }
        hashMap[num] = index
    }
    return []
}
//Time Complexity O(n) - The algorithm iterates through the array once, resulting in linear time complexity.
//Space Complexity O(n) - The algorithm uses a hash map to store the indices of the

